### PR TITLE
Enable HLO elementwise ops for int4 types

### PR DESCRIPTION
**Pull request by sergey-kozub**
Allow more HLO ops on sub-byte types: all elementwise ops and opt-barrier.

### CODE REVIEW

File: xla/service/cpu_gpu_shape_verifier.cc [hunk 1]
```diff
@@ -75,7 +76,9 @@ absl::Status VerifyS4U4Usage(HloInstruction* instruction) {
     case HloOpcode::kDynamicUpdateSlice:
     case HloOpcode::kFusion:
     case HloOpcode::kGetTupleElement:
+    case HloOpcode::kOptimizationBarrier:
     case HloOpcode::kParameter:
+    case HloOpcode::kSelect:
```
Comment (bchetioui) [inline xla/service/cpu_gpu_shape_verifier.cc:81]: Why is `select` here? It's an elementwise, right?

**Review by bchetioui — CHANGES_REQUESTED** [review]
Do we know that we want to support all elementwise ops on `int4`? What's the support plan here---I imagine that codegen will currently fail for all these ops, right?

File: xla/service/cpu_gpu_shape_verifier.cc (hunk 1 above)
Comment (sergey-kozub) [reply xla/service/cpu_gpu_shape_verifier.cc]: Indeed, same as kBitcastConvert. Removed both.

**Comment by sergey-kozub** [comment]
> What's the support plan here---I imagine that codegen will currently fail for all these ops, right?

Quite the opposite - this all works today (after this PR applied), I see no reason to keep this disabled.

**Comment by bchetioui** [comment]
> > What's the support plan here---I imagine that codegen will currently fail for all these ops, right?
> 
> Quite the opposite - this all works today (after this PR applied), I see no reason to keep this disabled.

Awesome! Are there codegen tests already enabled?

**Comment by sergey-kozub** [comment]
> Are there codegen tests already enabled?

I cannot find any tests, even though there is int4 specific code in e.g. "lower_tensors.cc"
I'll add a few basic tests to this PR that compare vs the interpreter.

**Comment by sergey-kozub** [comment]
Added a test, PTAL.

File: xla/tests/int4_test.cc [hunk 2]
```diff
@@ -157,6 +157,23 @@ XLA_TEST_F(HloTestBase, HorizontalLoopFusion) {
   EXPECT_TRUE(RunAndCompare(hlo_text, std::nullopt));
 }
 
+XLA_TEST_F(HloTestBase, ElementwiseOps) {
```
Comment (bchetioui) [inline xla/tests/int4_test.cc:160]: Could we make an exhaustive test that tests each and every unary, binary, and ternary elementwise op?

We could have three parametrized test suites.

**Review by bchetioui — CHANGES_REQUESTED** [review]
PTAL at the comment :)

File: xla/tests/int4_test.cc (hunk 2 above)
Comment (sergey-kozub) [reply xla/tests/int4_test.cc]: Added exhaustive tests for ops that support integer arguments.

File: xla/tests/int4_test.cc [hunk 3]
```diff
@@ -186,5 +190,161 @@ XLA_TEST_F(HloTestBaseWithAlgsimpDisabled, TwoDots) {
 )";
   EXPECT_TRUE(RunAndCompare(hlo_text, std::nullopt));
 }
+
+class ElementwiseTest : public HloTestBase,
+                        public ::testing::WithParamInterface<
+                            std::tuple<HloOpcode, PrimitiveType>> {
+ public:
+  static std::vector<HloOpcode> GetOpcodesWithArity(int arity) {
```
Comment (bchetioui) [inline xla/tests/int4_test.cc:198]: `GetElementwiseOpcodesWithIntSupportWithArity`?

The name of the function right now is not telling the whole story.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc [hunk 4]
```diff
@@ -162,26 +162,6 @@ TEST_F(TritonGemmTest, RejectDotInt4HLO) {
               StatusIs(tsl::error::INVALID_ARGUMENT));
 }
 
-TEST_F(TritonGemmTest, RejectInt4NegatePlusConvertHLO) {
```
Comment (bchetioui) [inline xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc:165]: What's the expectation here, that it's fused?

If so, should we check correctness of the lowering?

**Review by bchetioui — CHANGES_REQUESTED** [review]
Thanks Sergey, just a few more things related to testing.

File: xla/tests/int4_test.cc (hunk 3 above)
Comment (sergey-kozub) [reply xla/tests/int4_test.cc:198]: Updated the name as you suggest.

Note that the function name usually isn't telling the whole story, the function body does.

File: xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc (hunk 4 above)
Comment (sergey-kozub) [reply xla/service/gpu/fusions/triton/triton_fusion_emitter_device_legacy_test.cc:165]: Restored the tests, removed "Reject" from the test name.

File: xla/tests/int4_test.cc (hunk 3 above)
Comment (bchetioui) [reply xla/tests/int4_test.cc]: The function body shouldn't introduce behaviours that are not clear from the name + the docstring. In this case, the behaviour was quite different from what we could expect from just looking at the prototype---that is an antipattern.

**Review by bchetioui — APPROVED** [review]
Thank you Sergey!

**Pull request closed without merging**